File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    double binPow(double x, long n){
        if(n == 0)
            return 1;
        
        // handle negative exponent 
        if(n < 0){
            x = 1/x;
            n = -n;
        }

        // double half = binPow(x,n/2);
        
        if(n % 2 == 0){
            double temp = binPow(x, n/2);
            return temp * temp;
        }   
            
        else
            return x * binPow(x, n - 1);
    }
    double myPow(double x, int n) {
        //handle int overflow for exponent
        return binPow(x,(long)n);
    }
};

File: 0050-powx-n/0050-powx-n.java
class Solution {
    public double solve(double x, long n){
        if(n == 0)
            return 1.0;
        double half = solve(x, n / 2);
        if(n % 2 == 0)
            return half * half;
        return x * half * half;
    }
    public double myPow(double x, int n) {
        long N = n;
        if(N < 0)
            return solve(1/x, -N);
        return solve(x,N);
    }
}